test_contextual_lenses_learning.py

```diff
@@ -0,0 +1,629 @@
+            **encoder_fn_kwargs):
+    # Encode indices using embeddings
+    # Apply lensing operation
+    # Predict value
+
+    if encoder_fn is not None:
+        x = encoder_fn(x, **encoder_fn_kwargs)
+
+    rep = reduce_fn(x, linear_rep_size)
+
+    out = nn.Dense(rep,
+                   1,
+                   kernel_init=nn.initializers.xavier_uniform(),
+                   bias_init=nn.initializers.normal(stddev=1e-6))
+    
+    return out
+
+def create_test_representation_model(encoder_fn, reduce_fn, seq_len, rep_size,
+                                     linear_rep_size=256,
+                                     key=random.PRNGKey(0),
+                                **encoder_fn_kwargs):
+  module = TestRepresentationModel.partial(encoder_fn=encoder_fn, 
+                                       reduce_fn=reduce_fn,
+                                       rep_size=rep_size,
+                                       seq_len=seq_len,
+                                       linear_rep_size=linear_rep_size,
+                                       **encoder_fn_kwargs)
+  _, initial_params = TestRepresentationModel.init_by_shape(key,
+                                                     input_specs=[((1, seq_len, rep_size), jnp.float32)],
+                                                     encoder_fn=encoder_fn,
+                                                     reduce_fn=reduce_fn,
+                                                     rep_size=rep_size, 
+                                                     seq_len=seq_len,
+                                                     linear_rep_size=linear_rep_size,
+                                                     **encoder_fn_kwargs)
+  model = nn.Model(module, initial_params)
+  return model
+
+def create_optimizer(model, learning_rate, weight_decay):
+  optimizer_def = optim.Adam(learning_rate=learning_rate, weight_decay=weight_decay)
+  optimizer = optimizer_def.create(model)
+  return optimizer
+
+@jax.jit
+def train_step(optimizer, X, Y):
+  def loss_fn(model):
+    Y_hat = model(X)
+    Y_hat = jnp.squeeze(Y_hat, axis=1)
+    loss = jnp.mean(jnp.square(Y-Y_hat))
+    return loss
+  grad_fn = jax.value_and_grad(loss_fn)
+  _, grad = grad_fn(optimizer.target)
+  optimizer = optimizer.apply_gradient(grad)
+  return optimizer
+
+"""# Tests
+
+## Data
+"""
+
+bs = 5
+seq_len = 4
+rep_size = 10
+input_data = jnp.array(np.random.normal(scale=5, size=(bs, seq_len, rep_size)))
+output_data = jnp.array(np.random.normal(size=(bs)))
+
+loss_threshold = 1e-4
+
+def compute_train_loss(model, input_data, output_data, lr=1e-3):
+  
+  optimizer = create_optimizer(model, learning_rate=lr, weight_decay=0.)
+  
+  epochs = 150
+  for epoch in range(epochs):
+    optimizer = train_step(optimizer, input_data, output_data)
+
+  preds = jnp.squeeze(optimizer.target(input_data), axis=1)
+
+  train_loss = jnp.mean((preds-output_data)**2)
+
+  return train_loss
+
+"""## No encoding"""
+
+def test_mean():
```

Comment:
everything above here should be in the library code (util.py/train.py and contextual_lenses.py). Everything below here should be the test. It needs to be wrapped in a test class. I'd using absl.testing.TestCase.